conllu/models.py

```diff
@@ -19,70 +35,71 @@ class Token(OrderedDict):
         "xpos": "xpostag",
     }
 
-    def get(self, key, *args, **kwargs):
+    def get(self, key: str, default: T.Optional[T.Any] = None) -> T.Any:
         if key not in self and key in self.MAPPING:
             key = self.MAPPING[key]
 
-        return super(Token, self).get(key, *args, **kwargs)
+        return super(Token, self).get(key, default)
 
-    def __missing__(self, key):
+    def __missing__(self, key: str) -> T.Any:
         if key in self.MAPPING:
             return self[self.MAPPING[key]]
 
         raise KeyError("'" + key + "'")
 
-class TokenList(list):
-    metadata = None
+class TokenList(T.List[Token]):
+
+    metadata: T.Optional[_MetadataType] = None
 
-    def __init__(self, tokens, metadata=None):
+    def __init__(self, tokens: T.Iterable[Token], metadata: T.Optional[_MetadataType] = None):
         super(TokenList, self).__init__(tokens)
         if not isinstance(tokens, list):
             raise ParseException("Can't create TokenList, tokens is not a list.")
 
         self.metadata = metadata
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return 'TokenList<' + ', '.join(token['form'] for token in self) + '>'
 
-    def __eq__(self, other):
+    def __eq__(self, other: T.Any) -> bool:
         return super(TokenList, self).__eq__(other) \
             and (not hasattr(other, 'metadata') or self.metadata == other.metadata)
 
-    def __ne__(self, other):
+    def __ne__(self, other: T.Any) -> bool:
         return not self == other
 
-    def clear(self):
+    def clear(self) -> None:
         self[:] = []  # Supported in Python 2 and 3, unlike clear()
         self.metadata = None
 
-    def copy(self):
+    def copy(self) -> 'TokenList':
         tokens_copy = self[:]  # Supported in Python 2 and 3, unlike copy()
         return TokenList(tokens_copy, self.metadata)
 
-    def extend(self, iterable):
+    def extend(self, iterable: T.Iterable[Token]) -> None:
         super(TokenList, self).extend(iterable)
-        if hasattr(iterable, 'metadata'):
-            if hasattr(self.metadata, '__add__') and hasattr(iterable.metadata, '__add__'):
+        if isinstance(iterable, TokenList):
+            if isinstance(self.metadata, Addable) and isinstance(iterable.metadata, Addable):
```

Comment:
I introduced `Addable` "type checkable" way of checking if `self.metadata` has the `__add__` method. I actually wasn't clear on what exactly `__add__` is supposed to distinguish between(ie, what other types are valid types for `self.metadata`? list? tuple? str? All have `.__add__`). I think refactoring the code to be more explicit would make it clearer for future contributors.

Reply:
It's supposed to mean (and is only tested for) "list of dicts". See the test_extend in tests/test_models.py. As long as that works, any simplifications is OK. If I understand things correctly this also means we can skip the typing_extensions dependency?